File: frontend/src/types/peer-review.ts
// Mirrors app/models/enum.py::ReviewStatus.
// NOTE: nếu dự án đã có enum này trong '@/types/statuses', hãy xóa định nghĩa
// dưới đây và import từ đó thay vì khai báo lại.
export enum ReviewStatus {
    PENDING = 'PENDING',
    COMPLETED = 'COMPLETED',
    SKIPPED = 'SKIPPED',
}

// ---------------------------------------------------------------------------
// Danh sách bài được giao cho học viên (reviewer) chấm chéo
// ---------------------------------------------------------------------------
export interface MyAssignment {
    assignment_id: string;
    quiz_id: string;
    submission_id: string;
    status: ReviewStatus;
    final_score_given: number | null;
    assigned_at: string;
    completed_at: string | null;
}

// ---------------------------------------------------------------------------
// Chi tiết 1 lượt chấm: câu trả lời tự luận cần chấm kèm rubric tương ứng
// ---------------------------------------------------------------------------
export interface RubricCriteria {
    criteria_id: string;
    title: string;
    description: string | null;
    percentage: number;
}

export interface EssayAnswer {
    question_id: string;
    question_title: string;
    body_content: string;
    max_points: number;

    essay_answer_text: string | null;
    graph_image_url: string | null;
    graph_json_data: string | null;
    video_trigger_seconds: number | null;

    rubric_criterias: RubricCriteria[];
}

export interface AssignmentDetail extends MyAssignment {
    quiz_title: string;
    general_comment: string | null;
    answers: EssayAnswer[];
}

// ---------------------------------------------------------------------------
// Nộp kết quả chấm chéo
// ---------------------------------------------------------------------------
export interface EvaluationItemPayload {
    criteria_id: string;
    score: number;
    feedback?: string | null;
}

export interface SubmitEvaluationPayload {
    evaluations: EvaluationItemPayload[];
    general_comment?: string | null;
}

export interface EvaluationItem {
    evaluation_id: string;
    criteria_id: string;
    score: number;
    feedback: string | null;
}

export interface SubmitEvaluationResponse {
    assignment_id: string;
    status: ReviewStatus;
    final_score_given: number;
    general_comment: string | null;
    completed_at: string;
    evaluations: EvaluationItem[];

    // Trạng thái tổng hợp phía submission sau khi lượt chấm này được ghi nhận
    submission_fully_reviewed: boolean;
    submission_peer_avg_score: number | null;
    submission_is_discrepant: boolean | null;
}

// ---------------------------------------------------------------------------
// [Giảng viên] Bài nộp chấm chéo bị lệch điểm, cần chấm lại
// ---------------------------------------------------------------------------
export interface QuizPeerReviewInfo {
    quiz_id: string;
    title: string;
    is_peer_review: boolean;
}

export interface SubmissionListItem {
    submission_id: string;
    quiz_id: string;
    user_id: string;
    attempt_number: number;
    status: 'IN_PROGRESS' | 'SUBMITTED' | 'GRADED';
    is_peer_review: boolean;
    is_discrepant: boolean;
    peer_avg_score: number | null;
    completed_review_count: number;
    total_score: number | null;
    is_passed: boolean | null;
    started_at: string;
    submitted_at: string | null;
}

File: backend/service/quiz_exam_service/app/schemas/peer_review.py
import uuid
from uuid import UUID
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.enum import ReviewStatus, SubmissionStatus


# ---------------------------------------------------------------------------
# Danh sách bài được giao cho học viên (reviewer) chấm chéo
# ---------------------------------------------------------------------------
class MyAssignmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    assignment_id: UUID
    quiz_id: UUID
    submission_id: UUID
    status: ReviewStatus
    final_score_given: Optional[float] = None
    assigned_at: datetime
    completed_at: Optional[datetime] = None


# ---------------------------------------------------------------------------
# Chi tiết 1 lượt chấm: câu trả lời tự luận cần chấm kèm rubric tương ứng
# ---------------------------------------------------------------------------
class RubricCriteriaOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    criteria_id: UUID
    title: str
    description: Optional[str] = None
    percentage: float


class EssayAnswerOut(BaseModel):
    """Câu trả lời tự luận (ESSAY) của bài nộp, kèm rubric của câu hỏi đó."""

    question_id: UUID
    question_title: str
    body_content: str
    max_points: float

    essay_answer_text: Optional[str] = None
    graph_image_url: Optional[str] = None
    graph_json_data: Optional[str] = None
    video_trigger_seconds: Optional[int] = None

    rubric_criterias: List[RubricCriteriaOut] = Field(default_factory=list)


class AssignmentDetailOut(MyAssignmentOut):
    quiz_title: str
    general_comment: Optional[str] = None
    answers: List[EssayAnswerOut] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Nộp kết quả chấm chéo
# ---------------------------------------------------------------------------
class EvaluationItemIn(BaseModel):
    criteria_id: UUID
    score: float = Field(ge=0, description="Điểm cho riêng tiêu chí này")
    feedback: Optional[str] = None

    @field_validator("score")
    @classmethod
    def validate_score(cls, v: float) -> float:
        if v < 0:
            raise ValueError("Điểm không được nhỏ hơn 0")
        return v


class SubmitEvaluationIn(BaseModel):
    evaluations: List[EvaluationItemIn] = Field(min_length=1)
    general_comment: Optional[str] = None


class EvaluationItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    evaluation_id: UUID
    criteria_id: UUID
    score: float
    feedback: Optional[str] = None


class SubmitEvaluationOut(BaseModel):
    assignment_id: UUID
    status: ReviewStatus
    final_score_given: float
    general_comment: Optional[str] = None
    completed_at: datetime
    evaluations: List[EvaluationItemOut]

    # Trạng thái tổng hợp phía submission sau khi lượt chấm này được ghi nhận
    submission_fully_reviewed: bool
    submission_peer_avg_score: Optional[float] = None
    submission_is_discrepant: Optional[bool] = None


# ---------------------------------------------------------------------------
# [Giảng viên] Danh sách bài nộp theo cờ is_peer_review (2 mục: bài nộp thường / chấm chéo)
# ---------------------------------------------------------------------------
class QuizPeerReviewInfoOut(BaseModel):
    """Thông tin tối thiểu để FE biết đề thi có bật chấm chéo hay không, dùng để
    quyết định có hiển thị 2 mục (bài nộp thường / chấm chéo) hay không."""

    model_config = ConfigDict(from_attributes=True)

    quiz_id: UUID
    title: str
    is_peer_review: bool


class SubmissionListItemOut(BaseModel):
    """1 lượt nộp bài, dùng cho cả 2 mục:
    - is_peer_review=False -> mục 'Bài nộp thường'.
    - is_peer_review=True  -> mục 'Chấm chéo' (gồm cả bài đã tự động chốt điểm
      lẫn bài lệch điểm cần chấm lại, phân biệt qua is_discrepant)."""

    model_config = ConfigDict(from_attributes=True)

    submission_id: UUID
    quiz_id: UUID
    user_id: UUID
    attempt_number: int
    status: SubmissionStatus
    is_peer_review: bool
    is_discrepant: bool
    peer_avg_score: Optional[float] = None
    completed_review_count: int
    total_score: Optional[float] = None
    is_passed: Optional[bool] = None
    started_at: datetime
    submitted_at: Optional[datetime] = None
